README.md
# progressable_button

A progressable button using streams

CHANGELOG.md
## [1.0.0]

- Initial release

## [1.0.1]

- Code refactor

## [1.0.2]

- Bool verification fix

## [1.1.0]

- Possibility to add a gradient as background

## [1.1.1]

- Fix to force all colors as gradients in case there is a single one gradient in constructor

## [1.1.2]

- Fix to force all colors as gradients in case there is a single one gradient in constructor

## [1.1.3]

- Fix to force all colors as gradients in case there is a single one gradient in constructor

## [1.1.4]

- Initial enabled value in case the enabled stream is a rxdart's BehaviourSubject

lib/progress_button.dart
library progress_button;

import 'dart:async';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math_64.dart' as Vectors;
import 'package:vibrate/vibrate.dart';

class ProgressTextButton extends StatelessWidget {
  final String text;
  final TextStyle textStyle;

  final VoidCallback onClick;

  final Stream<bool> loadingStream;
  final Stream<bool> enabledStream;
  final Stream<Exception> errorStream;

  final double expandedSize;
  final bool big;

  final EdgeInsets margin;
  final EdgeInsets loadingMargin;
  final EdgeInsets padding;

  final BorderRadius idleBorderRadius;
  final BorderRadius loadingBorderRadius;

  final Widget loadingChild;

  final Color enabledColor;
  final Color disabledColor;
  final Color errorColor;

  final List<BoxShadow> shadow;

  ProgressTextButton({
    this.loadingStream,
    this.enabledStream,
    this.errorStream,
    this.onClick,
    this.text,
    this.textStyle,
    this.expandedSize,
    this.loadingMargin,
    this.idleBorderRadius,
    this.loadingBorderRadius,
    this.padding = const EdgeInsets.symmetric(
      vertical: 10,
      horizontal: 20,
    ),
    this.big,
    this.margin,
    this.enabledColor,
    this.loadingChild,
    this.disabledColor,
    this.errorColor,
    this.shadow,
  });

  @override
  Widget build(BuildContext context) {
    return ProgressButton(
      padding: padding,
      errorStream: errorStream,
      onClick: onClick,
      enabledColor: enabledColor,
      enabledStream: enabledStream,
      loadingStream: loadingStream,
      expandedSize: expandedSize,
      idleBorderRadius: idleBorderRadius,
      loadingMargin: loadingMargin,
      margin: margin,
      big: big,
      loadingChild: loadingChild,
      enabledChild: Text(
        text,
        style: textStyle != null
            ? textStyle
            : big
                ? Theme.of(context).textTheme.body1
                : Theme.of(context).textTheme.body2,
        textAlign: TextAlign.center,
      ),
    );
  }
}

class ProgressButton extends StatefulWidget {
  final Widget enabledChild;
  final Widget loadingChild;

  final VoidCallback onClick;

  final Stream<bool> loadingStream;
  final Stream<bool> enabledStream;
  final Stream<Exception> errorStream;

  final double expandedSize;
  final bool big;

  final EdgeInsets margin;
  final EdgeInsets loadingMargin;
  final EdgeInsets padding;

  final Color enabledColor;
  final Color disabledColor;
  final Color errorColor;

  final BorderRadius idleBorderRadius;
  final BorderRadius loadingBorderRadius;

  final List<BoxShadow> shadow;

  ProgressButton({
    this.loadingStream,
    this.enabledStream,
    this.errorStream,
    this.onClick,
    this.expandedSize,
    @required this.enabledChild,
    this.loadingChild,
    bool big,
    List<BoxShadow> shadow,
    Color enabledColor,
    Color errorColor,
    Color disabledColor,
    BorderRadius idleBorderRadius,
    BorderRadius loadingBorderRadius,
    EdgeInsets padding,
    EdgeInsets loadingMargin,
    EdgeInsets margin,
  })  : this.padding =
            padding ?? ProgressButtonDefaultConfiguration.idlePadding,
        this.loadingMargin =
            loadingMargin ?? ProgressButtonDefaultConfiguration.loadingMargin,
        this.margin = margin ?? ProgressButtonDefaultConfiguration.idleMargin,
        this.loadingBorderRadius = loadingBorderRadius ??
            ProgressButtonDefaultConfiguration.loadingBorderRadius,
        this.idleBorderRadius = idleBorderRadius ??
            ProgressButtonDefaultConfiguration.idleBorderRadius,
        this.enabledColor =
            enabledColor ?? ProgressButtonDefaultConfiguration.enabledColor,
        this.errorColor =
            errorColor ?? ProgressButtonDefaultConfiguration.errorColor,
        this.disabledColor =
            disabledColor ?? ProgressButtonDefaultConfiguration.disabledColor,
        this.shadow = shadow ?? ProgressButtonDefaultConfiguration.shadow,
        this.big = big ?? true;

  @override
  _ProgressButtonState createState() => _ProgressButtonState();
}

class _ProgressButtonState extends State<ProgressButton>
    with TickerProviderStateMixin {
  StreamSubscription _errorSubscription;
  StreamSubscription _loadingSubscription;
  StreamSubscription _enabledSubscription;

  AnimationController _errorAnimationController;

  bool _loading = false;
  bool _enabled = true;
  bool _inErrorAnimation = false;

  @override
  void initState() {
    super.initState();

    _errorAnimationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 500),
    )..addStatusListener(
        (status) {
          switch (status) {
            case AnimationStatus.dismissed:
              setState(() => _inErrorAnimation = false);
              break;
            case AnimationStatus.forward:
              setState(() => _inErrorAnimation = true);
              break;
            case AnimationStatus.reverse:
              setState(() => _inErrorAnimation = true);
              break;
            case AnimationStatus.completed:
              setState(() => _inErrorAnimation = false);
              break;
          }
        },
      );

    _errorSubscription = widget.errorStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (_) => _startShake(),
        );

    _loadingSubscription = widget.loadingStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (loading) => setState(() => _loading = loading),
        );

    _enabledSubscription = widget.enabledStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (enabled) => setState(() => _enabled = enabled),
        );
  }

  void _startShake() async {
    Vibrate.vibrate();

    _errorAnimationController.forward(
        from: _errorAnimationController.lowerBound);
  }

  Vectors.Vector3 _lerpShake() {
    double progress = _errorAnimationController.value;
    double offset = sin(progress * pi * 10.0);
    return Vectors.Vector3(offset * 8, 0.0, 0.0);
  }

  @override
  Widget build(BuildContext context) {
    return IgnorePointer(
      ignoring: _inErrorAnimation || !_enabled || _loading,
      child: Container(
        margin: _loading ? widget.loadingMargin : widget.margin,
        child: AnimatedBuilder(
          animation: _errorAnimationController,
          builder: (context, child) {
            return Transform(
              transform: Matrix4.translation(_lerpShake()),
              child: child,
            );
          },
          child: AnimatedContainer(
            duration: const Duration(milliseconds: 300),
            curve: Curves.ease,
            decoration: BoxDecoration(
              color: _backgroundColor,
              borderRadius: _borderRadius,
              boxShadow: widget.shadow,
            ),
            child: Stack(
              children: <Widget>[
                AnimatedSize(
                  vsync: this,
                  duration: const Duration(milliseconds: 300),
                  curve: Curves.ease,
                  child: _child,
                ),
                Positioned(
                  right: 0,
                  left: 0,
                  top: 0,
                  bottom: 0,
                  child: Material(
                    color: Colors.transparent,
                    child: InkWell(
                      onTap: widget.onClick,
                      borderRadius: _borderRadius,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget get _child {
    if (_loading) return widget.loadingChild ?? _buildDefaultLoadingWidget();

    return Container(
      padding: widget.padding,
      width: widget.expandedSize,
      child: widget.enabledChild,
    );
  }

  BorderRadius get _borderRadius {
    if (_loading) return widget.loadingBorderRadius;

    return widget.idleBorderRadius;
  }

  Color get _backgroundColor {
    if (_inErrorAnimation)
      return widget.errorColor ?? Theme.of(context).errorColor;

    if (_loading || !_enabled)
      return widget.disabledColor ?? Theme.of(context).disabledColor;

    return widget.enabledColor ?? Theme.of(context).primaryColor;
  }

  Widget _buildDefaultLoadingWidget() {
    return Container(
      padding: EdgeInsets.all(widget.big ? 12 : 8),
      child: SizedBox(
        height: widget.big ? 24.0 : 16.0,
        width: widget.big ? 24.0 : 16.0,
        child: const CircularProgressIndicator(
          valueColor: const AlwaysStoppedAnimation<Color>(
            Colors.white,
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _errorSubscription?.cancel();
    _loadingSubscription?.cancel();
    _enabledSubscription?.cancel();

    _errorAnimationController.dispose();

    super.dispose();
  }
}

class ProgressButtonDefaultConfiguration {
  static EdgeInsets idleMargin = EdgeInsets.zero;
  static EdgeInsets loadingMargin = EdgeInsets.zero;

  static EdgeInsets idlePadding = const EdgeInsets.symmetric(vertical: 10);

  static BorderRadius idleBorderRadius = BorderRadius.circular(10);
  static BorderRadius loadingBorderRadius = BorderRadius.circular(100);

  static Color enabledColor;
  static Color errorColor;
  static Color disabledColor;

  static List<BoxShadow> shadow = [
    BoxShadow(
      color: Colors.grey.withOpacity(0.5),
      blurRadius: 5.0,
      spreadRadius: 0.25,
      offset: Offset(0, 3),
    ),
  ];
}


lib/progressable_button.dart
library progressable_button;

import 'dart:async';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:rxdart/rxdart.dart';
import 'package:vector_math/vector_math_64.dart' as Vectors;
import 'package:vibrate/vibrate.dart';

class ProgressableTextButton extends StatelessWidget {
  final String text;
  final TextStyle textStyle;

  final VoidCallback onClick;

  final Stream<bool> loadingStream;
  final Stream<bool> enabledStream;
  final Stream<Exception> errorStream;

  final double expandedSize;
  final bool big;

  final EdgeInsets margin;
  final EdgeInsets loadingMargin;
  final EdgeInsets padding;

  final BorderRadius idleBorderRadius;
  final BorderRadius loadingBorderRadius;

  final Widget loadingChild;

  final Color enabledColor;
  final Color disabledColor;
  final Color errorColor;

  final Gradient enabledGradient;
  final Gradient disabledGradient;
  final Gradient errorGradient;

  final List<BoxShadow> shadow;

  ProgressableTextButton({
    this.loadingStream,
    this.enabledStream,
    this.errorStream,
    this.onClick,
    this.text,
    this.textStyle,
    this.expandedSize,
    this.loadingMargin,
    this.idleBorderRadius,
    this.loadingBorderRadius,
    this.padding = const EdgeInsets.symmetric(
      vertical: 10,
      horizontal: 20,
    ),
    bool big,
    this.margin,
    this.enabledColor,
    this.loadingChild,
    this.disabledColor,
    this.errorColor,
    this.shadow,
    this.enabledGradient,
    this.disabledGradient,
    this.errorGradient,
  }) : this.big = big ?? true;

  @override
  Widget build(BuildContext context) {
    return ProgressableButton(
      padding: padding,
      errorStream: errorStream,
      onClick: onClick,
      enabledColor: enabledColor,
      enabledStream: enabledStream,
      loadingStream: loadingStream,
      expandedSize: expandedSize,
      idleBorderRadius: idleBorderRadius,
      loadingMargin: loadingMargin,
      margin: margin,
      big: big,
      loadingChild: loadingChild,
      enabledChild: Text(
        text,
        style: textStyle != null
            ? textStyle
            : big
                ? Theme.of(context).textTheme.body1
                : Theme.of(context).textTheme.body2,
        textAlign: TextAlign.center,
      ),
      shadow: shadow,
      loadingBorderRadius: loadingBorderRadius,
      errorColor: errorColor,
      disabledColor: disabledColor,
      disabledGradient: disabledGradient,
      enabledGradient: enabledGradient,
      errorGradient: errorGradient,
    );
  }
}

class ProgressableButton extends StatefulWidget {
  final Widget enabledChild;
  final Widget loadingChild;

  final VoidCallback onClick;

  final Stream<bool> loadingStream;
  final Stream<bool> enabledStream;
  final Stream<Exception> errorStream;

  final double expandedSize;
  final bool big;

  final EdgeInsets margin;
  final EdgeInsets loadingMargin;
  final EdgeInsets padding;

  final Color enabledColor;
  final Color disabledColor;
  final Color errorColor;

  final Gradient enabledGradient;
  final Gradient disabledGradient;
  final Gradient errorGradient;

  final BorderRadius idleBorderRadius;
  final BorderRadius loadingBorderRadius;

  final List<BoxShadow> shadow;

  ProgressableButton({
    this.loadingStream,
    this.enabledStream,
    this.errorStream,
    this.onClick,
    this.expandedSize,
    @required this.enabledChild,
    this.loadingChild,
    this.enabledGradient,
    this.disabledGradient,
    this.errorGradient,
    bool big,
    List<BoxShadow> shadow,
    Color enabledColor,
    Color errorColor,
    Color disabledColor,
    BorderRadius idleBorderRadius,
    BorderRadius loadingBorderRadius,
    EdgeInsets padding,
    EdgeInsets loadingMargin,
    EdgeInsets margin,
  })  : this.padding =
            padding ?? ProgressableButtonDefaultConfiguration.idlePadding,
        this.loadingMargin = loadingMargin ??
            ProgressableButtonDefaultConfiguration.loadingMargin,
        this.margin =
            margin ?? ProgressableButtonDefaultConfiguration.idleMargin,
        this.loadingBorderRadius = loadingBorderRadius ??
            ProgressableButtonDefaultConfiguration.loadingBorderRadius,
        this.idleBorderRadius = idleBorderRadius ??
            ProgressableButtonDefaultConfiguration.idleBorderRadius,
        this.enabledColor =
            enabledColor ?? ProgressableButtonDefaultConfiguration.enabledColor,
        this.errorColor =
            errorColor ?? ProgressableButtonDefaultConfiguration.errorColor,
        this.disabledColor = disabledColor ??
            ProgressableButtonDefaultConfiguration.disabledColor,
        this.shadow = shadow ?? ProgressableButtonDefaultConfiguration.shadow,
        this.big = big ?? true;

  @override
  _ProgressableButtonState createState() => _ProgressableButtonState();
}

class _ProgressableButtonState extends State<ProgressableButton>
    with TickerProviderStateMixin {
  StreamSubscription _errorSubscription;
  StreamSubscription _loadingSubscription;
  StreamSubscription _enabledSubscription;

  AnimationController _errorAnimationController;

  bool _loading = false;
  bool _enabled = true;
  bool _inErrorAnimation = false;

  @override
  void initState() {
    super.initState();

    if (widget.enabledStream is BehaviorSubject<bool>) {
      _enabled = (widget.enabledStream as BehaviorSubject<bool>).value;
    }

    _errorAnimationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 500),
    )..addStatusListener(
        (status) {
          switch (status) {
            case AnimationStatus.dismissed:
              setState(() => _inErrorAnimation = false);
              break;
            case AnimationStatus.forward:
              setState(() => _inErrorAnimation = true);
              break;
            case AnimationStatus.reverse:
              setState(() => _inErrorAnimation = true);
              break;
            case AnimationStatus.completed:
              setState(() => _inErrorAnimation = false);
              break;
          }
        },
      );

    _errorSubscription = widget.errorStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (_) => _startShake(),
        );

    _loadingSubscription = widget.loadingStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (loading) => setState(() => _loading = loading),
        );

    _enabledSubscription = widget.enabledStream
        ?.where(
          (event) => event != null,
        )
        ?.listen(
          (enabled) => setState(() => _enabled = enabled),
        );
  }

  void _startShake() async {
    Vibrate.vibrate();

    _errorAnimationController.forward(
        from: _errorAnimationController.lowerBound);
  }

  Vectors.Vector3 _lerpShake() {
    double progress = _errorAnimationController.value;
    double offset = sin(progress * pi * 10.0);
    return Vectors.Vector3(offset * 8, 0.0, 0.0);
  }

  @override
  Widget build(BuildContext context) {
    return IgnorePointer(
      ignoring: _inErrorAnimation || !_enabled || _loading,
      child: Container(
        margin: _loading ? widget.loadingMargin : widget.margin,
        child: AnimatedBuilder(
          animation: _errorAnimationController,
          builder: (context, child) {
            return Transform(
              transform: Matrix4.translation(_lerpShake()),
              child: child,
            );
          },
          child: AnimatedContainer(
            duration: const Duration(milliseconds: 300),
            curve: Curves.ease,
            decoration: BoxDecoration(
              gradient: _backgroundGradient,
              color: _backgroundColor,
              borderRadius: _borderRadius,
              boxShadow: widget.shadow,
            ),
            child: Stack(
              children: <Widget>[
                AnimatedSize(
                  vsync: this,
                  duration: const Duration(milliseconds: 300),
                  curve: Curves.ease,
                  child: _child,
                ),
                Positioned(
                  right: 0,
                  left: 0,
                  top: 0,
                  bottom: 0,
                  child: Material(
                    color: Colors.transparent,
                    child: InkWell(
                      onTap: widget.onClick,
                      borderRadius: _borderRadius,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget get _child {
    if (_loading) return widget.loadingChild ?? _buildDefaultLoadingWidget();

    return Container(
      padding: widget.padding,
      width: widget.expandedSize,
      child: widget.enabledChild,
    );
  }

  BorderRadius get _borderRadius {
    if (_loading) return widget.loadingBorderRadius;

    return widget.idleBorderRadius;
  }

  Gradient get _backgroundGradient {
    if (widget.enabledGradient != null ||
        widget.errorGradient != null ||
        widget.disabledGradient != null) {
      Gradient gradient;

      if (_inErrorAnimation)
        gradient = widget.errorGradient;
      else if (_loading || !_enabled)
        gradient = widget.disabledGradient;
      else
        gradient = widget.enabledGradient;

      return gradient ??
          LinearGradient(
            colors: [
              _backgroundColor,
              _backgroundColor,
            ],
          );
    }

    if (_inErrorAnimation) return widget.errorGradient;

    if (_loading || !_enabled) return widget.disabledGradient;

    return widget.enabledGradient;
  }

  Color get _backgroundColor {
    if (_inErrorAnimation)
      return widget.errorColor ?? Theme.of(context).errorColor;

    if (_loading || !_enabled)
      return widget.disabledColor ?? Theme.of(context).disabledColor;

    return widget.enabledColor ?? Theme.of(context).primaryColor;
  }

  Widget _buildDefaultLoadingWidget() {
    return Container(
      padding: EdgeInsets.all(widget.big ? 12 : 8),
      child: SizedBox(
        height: widget.big ? 24.0 : 16.0,
        width: widget.big ? 24.0 : 16.0,
        child: const CircularProgressIndicator(
          valueColor: const AlwaysStoppedAnimation<Color>(
            Colors.white,
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _errorSubscription?.cancel();
    _loadingSubscription?.cancel();
    _enabledSubscription?.cancel();

    _errorAnimationController.dispose();

    super.dispose();
  }
}

class ProgressableButtonDefaultConfiguration {
  static EdgeInsets idleMargin = EdgeInsets.zero;
  static EdgeInsets loadingMargin = EdgeInsets.zero;

  static EdgeInsets idlePadding = const EdgeInsets.symmetric(vertical: 10);

  static BorderRadius idleBorderRadius = BorderRadius.circular(10);
  static BorderRadius loadingBorderRadius = BorderRadius.circular(100);

  static Color enabledColor;
  static Color errorColor;
  static Color disabledColor;

  static List<BoxShadow> shadow = [
    BoxShadow(
      color: Colors.grey.withOpacity(0.5),
      blurRadius: 5.0,
      spreadRadius: 0.25,
      offset: Offset(0, 3),
    ),
  ];
}
